File: android-native/app/src/main/java/com/example/kpkn/screens/workout/WorkoutVoiceUi.kt
package com.example.kpkn.screens.workout

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.scaleIn
import androidx.compose.animation.scaleOut
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Mic
import androidx.compose.material.icons.filled.MicOff
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.SmallFloatingActionButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.zIndex
import com.example.kpkn.services.workout.VoicePipelineStage

@Composable
fun WorkoutVoiceFab(
    isEnabled: Boolean,
    voiceStage: VoicePipelineStage,
    onToggle: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val infiniteTransition = rememberInfiniteTransition(label = "voice_pulse")
    val pulseScale by infiniteTransition.animateFloat(
        initialValue = 1f,
        targetValue = 1.12f,
        animationSpec = infiniteRepeatable(
            animation = tween(800),
            repeatMode = RepeatMode.Reverse,
        ),
        label = "pulse_scale",
    )
    val pulseAlpha by infiniteTransition.animateFloat(
        initialValue = 0.7f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(600),
            repeatMode = RepeatMode.Reverse,
        ),
        label = "pulse_alpha",
    )

    val isListening = voiceStage == VoicePipelineStage.LISTENING
    val isProcessing = voiceStage == VoicePipelineStage.PROCESSING ||
            voiceStage == VoicePipelineStage.CONFIRM_WAIT ||
            voiceStage == VoicePipelineStage.TTS_SPEAKING

    val containerColor = when {
        isListening -> Color(0xFF4CAF50)
        isProcessing -> MaterialTheme.colorScheme.tertiary
        voiceStage == VoicePipelineStage.ERROR_RECOVERY -> Color(0xFFFFA000)
        isEnabled -> MaterialTheme.colorScheme.secondary
        else -> MaterialTheme.colorScheme.surfaceVariant
    }

    val icon = when {
        isEnabled -> Icons.Default.Mic
        else -> Icons.Default.MicOff
    }

    SmallFloatingActionButton(
        onClick = onToggle,
        modifier = modifier
            .zIndex(5f)
            .then(
                if (isListening) Modifier.scale(pulseScale).alpha(pulseAlpha)
                else Modifier
            ),
        containerColor = containerColor,
        contentColor = MaterialTheme.colorScheme.onPrimary,
        shape = CircleShape,
    ) {
        Icon(
            imageVector = icon,
            contentDescription = if (isEnabled) "Desactivar voz" else "Activar voz",
        )
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun WorkoutVoiceStatusBar(
    voiceStage: VoicePipelineStage,
    voicePartialText: String,
    voiceErrorMessage: String?,
    modifier: Modifier = Modifier,
) {
    val isVisible = voiceStage != VoicePipelineStage.DISABLED

    AnimatedVisibility(
        visible = isVisible,
        modifier = modifier.fillMaxWidth().zIndex(4f),
        enter = fadeIn(tween(150)) + scaleIn(tween(150)),
        exit = fadeOut(tween(200)) + scaleOut(tween(200)),
    ) {
        val (statusText, bgColor) = when (voiceStage) {
            VoicePipelineStage.LISTENING -> {
                if (voicePartialText.isNotBlank()) "Escuchando: \"$voicePartialText\"" to Color(0xFF4CAF50)
                else "Escuchando comandos..." to Color(0xFF4CAF50)
            }
            VoicePipelineStage.PROCESSING -> "Procesando..." to Color(0xFF795548)
            VoicePipelineStage.CONFIRM_WAIT -> "\"Sí\" para confirmar, \"No\" para cancelar" to Color(0xFF7CB342)
            VoicePipelineStage.TTS_SPEAKING -> "Hablando..." to Color(0xFF1565C0)
            VoicePipelineStage.ERROR_RECOVERY -> "Error, reintentando..." to Color(0xFFFFA000)
            VoicePipelineStage.DISABLED -> "" to Color.Transparent
        }

        if (statusText.isNotBlank()) {
            Surface(
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(0.dp),
                color = bgColor.copy(alpha = 0.92f),
                tonalElevation = 4.dp,
            ) {
                Column(
                    modifier = Modifier.padding(horizontal = 16.dp, vertical = 10.dp),
                ) {
                    Text(
                        text = statusText,
                        color = Color.White,
                        fontSize = 13.sp,
                        fontWeight = FontWeight.SemiBold,
                    )
                    if (voiceErrorMessage != null && voiceStage == VoicePipelineStage.ERROR_RECOVERY) {
                        Spacer(Modifier.height(2.dp))
                        Text(
                            text = voiceErrorMessage,
                            color = Color(0xFFFFCDD2),
                            fontSize = 11.sp,
                        )
                    }
                }
            }
        }
    }
}

File: ios-native/KPKNFit/KPKNFit/Presentation/Screens/Workout/WorkoutVoiceUi.swift
import SwiftUI

enum VoicePipelineStage: Equatable {
    case disabled
    case listening
    case processing
    case confirmWait
    case ttsSpeaking
    case errorRecovery
}

struct WorkoutVoiceFab: View {
    let isEnabled: Bool
    let voiceStage: VoicePipelineStage
    let onToggle: () -> Void

    @State private var pulseScale: CGFloat = 1.0
    @State private var pulseAlpha: CGFloat = 1.0

    var body: some View {
        let isListening = voiceStage == .listening
        let isProcessing = voiceStage == .processing || voiceStage == .confirmWait || voiceStage == .ttsSpeaking

        let containerColor: Color = {
            if isListening { return .green }
            if isProcessing { return .purple }
            if voiceStage == .errorRecovery { return .orange }
            if isEnabled { return Color(.systemGray2) }
            return Color(.systemGray5)
        }()

        Button(action: onToggle) {
            Image(systemName: isEnabled ? "mic.fill" : "mic.slash.fill")
                .font(.system(size: 16, weight: .semibold))
                .foregroundColor(.white)
                .frame(width: 40, height: 40)
                .background(containerColor)
                .clipShape(Circle())
        }
        .scaleEffect(isListening ? pulseScale : 1.0)
        .opacity(isListening ? pulseAlpha : 1.0)
        .onAppear {
            guard isListening else { return }
            withAnimation(.easeInOut(duration: 0.8).repeatForever(autoreverses: true)) {
                pulseScale = 1.12
                pulseAlpha = 0.7
            }
        }
    }
}

struct WorkoutVoiceStatusBar: View {
    let voiceStage: VoicePipelineStage
    let voicePartialText: String
    let voiceErrorMessage: String?

    var body: some View {
        if voiceStage != .disabled {
            let status = statusInfo
            if !status.text.isEmpty {
                VStack(spacing: 0) {
                    HStack {
                        Text(status.text)
                            .font(.system(size: 13, weight: .semibold))
                            .foregroundColor(.white)
                        Spacer()
                    }
                    .padding(.horizontal, 16)
                    .padding(.vertical, 10)
                    .background(status.color.opacity(0.92))

                    if let err = voiceErrorMessage, voiceStage == .errorRecovery {
                        HStack {
                            Text(err)
                                .font(.system(size: 11))
                                .foregroundColor(Color(red: 1, green: 0.8, blue: 0.84))
                            Spacer()
                        }
                        .padding(.horizontal, 16)
                        .padding(.bottom, 4)
                    }
                }
            }
        }
    }

    private var statusInfo: (text: String, color: Color) {
        switch voiceStage {
        case .listening:
            let partial = voicePartialText.trimmingCharacters(in: .whitespaces)
            return (partial.isEmpty ? "Escuchando comandos..." : "Escuchando: \"\(partial)\"", .green)
        case .processing: return ("Procesando...", Color(red: 0.47, green: 0.33, blue: 0.28))
        case .confirmWait: return ("\"Sí\" para confirmar, \"No\" para cancelar", Color(red: 0.49, green: 0.70, blue: 0.26))
        case .ttsSpeaking: return ("Hablando...", Color(red: 0.08, green: 0.40, blue: 0.75))
        case .errorRecovery: return ("Error, reintentando...", .orange)
        case .disabled: return ("", .clear)
        }
    }
}
